libs/extended/src/Network/HTTP/Client/Extended.hs

```diff
@@ -0,0 +1,101 @@
+
+import qualified STMContainers.Map as STMMap
+
+-- | Connection destination: just host and port.
+data Destination = Destination {-# UNPACK #-} !ShortByteString !Int
+    deriving (Eq, Show, Generic)
+
+instance Hashable Destination
+
+-- | Add a per-host hard connection limit. When there are more than N open connections to a
+-- host, any new connection to that host will immediately fail with 'PerHostConnectionLimit'.
+--
+-- Existing limits in 'Manager' are more fine-grained: for instance, those limits apply to raw
+-- and TLS connections separately. The limit set by 'addHardConnectionLimit' applies to all
+-- connections to the host/port, regardless of their type.
+addPerHostConnectionLimit :: MonadIO m => Int -> ManagerSettings -> m ManagerSettings
+addPerHostConnectionLimit limit managerSettings = liftIO $ do
+    -- Create a map counting active connections per host.
+    connCount :: STMMap.Map Destination Int <- STMMap.newIO
+
+    -- When a connection is created, increment a counter (and possibly throw an exception).
+    let onCreate :: Destination -> IO ()
+        onCreate dest@(Destination host_ port_) = do
+            success <- atomically $ do
+                current <- fromMaybe 0 <$> STMMap.lookup dest connCount
+                when (current < limit) $ STMMap.insert (current + 1) dest connCount
+                pure (current < limit)
+            unless success $ throwIO $
+                PerHostConnectionLimitReached (fromShort host_) port_
+
+    -- When a connection is destroyed, which is something 'Manager' should always do,
+    -- decrement the counter.
+    let onDestroy :: Destination -> IO ()
+        onDestroy dest = atomically $ do
+            current <- fromMaybe 0 <$> STMMap.lookup dest connCount
+            if current <= 1
+                then STMMap.delete dest connCount
+                else STMMap.insert (current - 1) dest connCount
+
+    -- Modify all connection-creating functions in the 'Manager' to perform 'onCreate' and
+    -- 'onDestroy'. Technically 'connectionClose' can be performed on an already closed
+    -- connection and we have to handle that gracefully (at least according to the
+    -- 'Connection' docs), but 'Manager' seems to close connections exactly once.
+    let managerRawConnection' = do
+            create <- managerRawConnection managerSettings
+            pure $ \addr host_ port_ -> do
+                let dest = Destination (fromString host_) port_
+                conn <- onCreate dest >> create addr host_ port_
+                pure (conn & connectionCloseL %~ (onDestroy dest >>))
+
+        managerTlsConnection' = do
+            create <- managerTlsConnection managerSettings
+            pure $ \addr host_ port_ -> do
+                let dest = Destination (fromString host_) port_
+                conn <- onCreate dest >> create addr host_ port_
+                pure (conn & connectionCloseL %~ (onDestroy dest >>))
+
```

Comment:
Minor nitpick: this block can be refactored since it's exactly the same in `managerRawConnection'` and `managerTlsConnection'`
```
let dest = Destination (fromString host_) port_
conn <- onCreate dest >> create addr host_ port_
pure (conn & connectionCloseL %~ (onDestroy dest >>))
```